analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/LoopDetector.cs

```diff
@@ -0,0 +1,92 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2024 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarAnalyzer.CFG.Roslyn;
+
+namespace SonarAnalyzer.SymbolicExecution.Roslyn;
+
+internal class LoopDetector
+{
+    private readonly HashSet<int> loopBlock;
+
+    public LoopDetector(ControlFlowGraph cfg)
+    {
+        loopBlock = DetectLoopBlockOrdinals(cfg).ToHashSet();
+    }
+
+    public bool IsInLoop(BasicBlock block) =>
+        loopBlock.Contains(block.Ordinal);
+
+    // Detects loops in the cfg using a modified DFS algorithm:
+    // Explore the cfg depth-first, keeping track of the path from the root to the current block.
+    // If the current block is already in the path, a loop is detected.
+    // Otherwise, if the last block in the path is part of an already detected loop that intersects with an earlier part of the path, merge the relevant sub-path with the loop.
+    // Due to the nature of DFS:
+    // If a block is processed already, current path has no loop and is not intersecting with already detected loops, the block is not part of any loop.
+    private static IEnumerable<int> DetectLoopBlockOrdinals(ControlFlowGraph cfg)
+    {
+        var processed = new HashSet<int>();
+        var loops = new List<HashSet<int>>();
+        var toProcess = new Stack<List<int>>();
+        toProcess.Push([1]);
+        while (toProcess.TryPop(out var path))
+        {
+            var last = path[path.Count - 1];
+            if (processed.Add(last))
+            {
+                ScheduleSuccessors(path, last);
+            }
+            else
+            {
+                var index = path.IndexOf(last);
+                if (index < path.Count - 1)         // is the block in the path twice? -> loop
+                {
+                    loops.Add(path.GetRange(index, path.Count - 1 - index).ToHashSet());    // equivalent to [index..^1]
```

Comment:
I find `[index..^1]` much easier to understand because it relies on an index instead of a count. I don't think the comments hurt anyone and it will make it easier to parse for some.

Reply:
We're not writing assembler that requires explaining syntax in comments. `path.GetRange(index, path.Count - 1 - index)` is pretty clear, and adding cryptic comment about `^1` that we don't intuitively understand makes the codebase worse, not better. 

Please remove it.

Reply:
It's funny that I find `path.GetRange(index, path.Count - 1 - index)` extremly cryptic and hard to understand while I find `[index..^1]` super clear and easy and it is the polar opposite in your case.